### PR TITLE
migrate PReLU to ATen

### LINKED ISSUE #10723: Segfault during backward when using PReLU
I'm trying to train a network using PReLU module, but I get Segfault during the backward. Here's a piece of code that reproduces the bug:

    gt = torch.rand(2,3,256,256)
    gt = torch.autograd.Variable(gt.cuda(async=True))
    input = torch.rand(2,134,256,256)
    input = torch.autograd.Variable(input.cuda())
    lossL1 = torch.nn.L1Loss()
    lossL1 = lossL1.cuda()

    net = nn.Sequential(nn.PReLU(), nn.Conv2d(134, 3, kernel_size=1, stride=1, bias=False)).cuda()

    output = net(input)

    loss = lossL1(output, gt)
    loss.backward()

In this example, my network just consists in a PReLU followed by a simple convolution. Note that if I switch the order of both modules, the Segfault doesn't occur, so it only bugs when the PReLU is the first layer.

Also note that if I don't use the GPU, the Segfault doesn't occur neither.

Rem: I tried with pytorch versions 0.4.0 and 0.4.1.

### PR DESCRIPTION
**Pull request by weiyangfb**
- fixes https://github.com/pytorch/pytorch/issues/10723
- migrate PReLU to ATen and deprecate legacy PReLU
- performance:

CPU with weight.numel() = 1
```
# === previous ===
>>> m = nn.PReLU()
>>> x = torch.randn(100, 100, 100, requires_grad=True)
>>> %timeit -r 100 y = m(x)
1000 loops, best of 100: 1.5 ms per loop

>>> y = m(x).sum()
>>> %timeit -r 100 y.backward(retain_graph=True)
10 loops, best of 100: 14.9 ms per loop

# === current ====
>>> m = nn.PReLU()
>>> x = torch.randn(100, 100, 100, requires_grad=True)
>>> %timeit -r 100 y = m(x)
1000 loops, best of 100: 179 µs per loop

>>> y = m(x).sum()
>>> %timeit -r 100 y.backward(retain_graph=True)
100 loops, best of 100: 378 µs per loop
```

CPU with weight.numel() = channels
```
# === previous ===
>>> m = nn.PReLU(100)
>>> x = torch.randn(100, 100, 100, requires_grad=True)
>>> %timeit -r 100 y = m(x)
1000 loops, best of 100: 351 µs per loop

>>> y = m(x).sum()
>>> %timeit -r 100 y.backward(retain_graph=True)
10 loops, best of 100: 7.14 ms per loop

# === current ====
>>> m = nn.PReLU(100)
>>> x = torch.randn(100, 100, 100, requires_grad=True)
>>> %timeit -r 100 y = m(x)
1000 loops, best of 100: 260 µs per loop

>>> y = m(x).sum()
>>> %timeit -r 100 y.backward(retain_graph=True)
100 loops, best of 100: 854 µs per loop
```

CUDA with weight.numel() = 1
```
# === previous ===
>>> m = nn.PReLU().cuda()
>>> x = torch.randn(100, 100, 100, requires_grad=True).cuda()
>>> %timeit -r 100 torch.cuda.synchronize(); y = m(x); torch.cuda.synchronize();
10000 loops, best of 100: 97.2 µs per loop

>>> y = m(x).sum()
>>> %timeit -r 100 torch.cuda.synchronize(); y.backward(retain_graph=True); torch.cuda.synchronize();
1000 loops, best of 100: 1.3 ms per loop

# === current ====
>>> m = nn.PReLU().cuda()
>>> x = torch.randn(100, 100, 100, requires_grad=True).cuda()
>>> %timeit -r 100 torch.cuda.synchronize(); y = m(x); torch.cuda.synchronize();
10000 loops, best of 100: 78.9 µs per loop

>>> y = m(x).sum()
>>> %timeit -r 100 torch.cuda.synchronize(); y.backward(retain_graph=True); torch.cuda.synchronize();
1000 loops, best of 100: 1.27 ms per loop
```

CUDA with weight.numel() = channel
```
# === previous ===
>>> m = nn.PReLU(100).cuda()
>>> x = torch.randn(100, 100, 100, requires_grad=True).cuda()
>>> %timeit -r 100 torch.cuda.synchronize(); y = m(x); torch.cuda.synchronize();
10000 loops, best of 100: 101 µs per loop

>>> y = m(x).sum()
>>> %timeit -r 100 torch.cuda.synchronize(); y.backward(retain_graph=True); torch.cuda.synchronize();
1000 loops, best of 100: 1.67 ms per loop

# === current ====
>>> m = nn.PReLU(100).cuda()
>>> x = torch.randn(100, 100, 100, requires_grad=True).cuda()
>>> %timeit -r 100 torch.cuda.synchronize(); y = m(x); torch.cuda.synchronize();
10000 loops, best of 100: 76.9 µs per loop

>>> y = m(x).sum()
>>> %timeit -r 100 torch.cuda.synchronize(); y.backward(retain_graph=True); torch.cuda.synchronize();
1000 loops, best of 100: 1.12 ms per loop
```

Updated runtimes after removing DEBUG flag when building PyTorch..

@ezyang @SsnL @zou3519  @soumith

### CODE REVIEW

**Comment by ezyang** [comment]
Not sure about the perf, but shouldn't you delete the old THNN implementation now?

**Comment by ezyang** [comment]
Can we get a billing of changes, i.e., what you changed when you did the port?

**Comment by fmassa** [comment]
What's the plan with `TensorIterator`? Is it going to be replacing `CPU_tensor_apply`?

**Comment by fmassa** [comment]
Should we instead try using `TensorIterator` here instead of the `CPU_tensor_apply`, or is it not yet there?

File: aten/src/ATen/native/Activation.cpp [hunk 1]
```diff
@@ -43,6 +43,205 @@ Tensor & rrelu_(Tensor & self, Scalar lower, Scalar upper, bool training, Genera
...
+
+  AT_CHECK(input.is_contiguous());
+  AT_CHECK(weight.is_contiguous());
+
+  int64_t weight_num = weight.numel();
+  Tensor result = at::empty_like(input);
+  auto strides = input.strides();
+
+  // case1: shared weight for all channels
+  if (weight_num == 1) {
```
Comment (weiyangfb) [inline aten/src/ATen/native/Activation.cpp]: @ezyang I mostly just copy&past from previous implementation, why there will be such a big performance gap?
https://github.com/pytorch/pytorch/blob/3da8d71d7de6f2eb3c9bbe395e38dc2fdf2784b0/aten/src/THNN/generic/PReLU.c#L5-L23

File: aten/src/ATen/native/Activation.cpp [hunk 2]
```diff
@@ -43,6 +43,205 @@ Tensor & rrelu_(Tensor & self, Scalar lower, Scalar upper, bool training, Genera
...
+  int64_t weight_num = weight.numel();
+  auto strides = input.strides();
+  auto dims = input.dim();
+
+  Tensor input_grad = at::empty_like(input);
+  Tensor weight_grad = at::empty_like(weight);
+  Tensor weight_grad_collector = at::empty_like(input);
+
+  // case1: shared parameter for all channels
+  if (weight_num == 1) {
```
Comment (weiyangfb) [inline aten/src/ATen/native/Activation.cpp]: CPU backward should also be similar to previous implementation:
https://github.com/pytorch/pytorch/blob/3da8d71d7de6f2eb3c9bbe395e38dc2fdf2784b0/aten/src/THNN/generic/PReLU.c#L71-L81

and 
https://github.com/pytorch/pytorch/blob/3da8d71d7de6f2eb3c9bbe395e38dc2fdf2784b0/aten/src/THNN/generic/PReLU.c#L146-L156

**Comment by weiyangfb** [comment]
@fmassa I wish I could use TensorIterator instead, but I haven't found a direct replacement to `CPU_tensor_apply *`, right now I just parallelized the CPU kernels to mitigate the performance issues.

File: aten/src/ATen/native/Activation.cpp [hunk 3]
```diff
@@ -43,6 +43,205 @@ Tensor & rrelu_(Tensor & self, Scalar lower, Scalar upper, bool training, Genera
...
+  auto input_data = input.data<scalar_t>();
+  auto weight_data = weight.data<scalar_t>();
+
+  // multiply values at each channel with weight[channel_index]
+  #pragma omp parallel for private(i,j,k) if (input.numel() > 10000)
+  for (i = 0; i < input_dim0_size; i++) {
+    for (j = 0; j < channel_size; j++) {
+      for (k = 0; k < input_stride1; k++) {
+        int64_t pos = i * input_stride0 + j * input_stride1 + k;
+        result_data[pos] = (input_data[pos] > 0) ? input_data[pos] : weight_data[j] * input_data[pos];
```
Comment (resistor) [inline aten/src/ATen/native/Activation.cpp]: This code can be made more efficient by communicating the the fast that speculation is acceptable to the compiler:

> scalar_t weight = weight_data[j];
> scalar_t input = input_data[pos];
> scalar_t scaled_output = weight * input;
> result_data[pos] = (input > 0) ? input : scaled_output;

You can go even farther down the rabbit hole by hoisting the computation of `weight` out a loop level, and by using __restrict__ and the aligned attributed as appropriate.  Maybe not worth the effort.

Comment (resistor) [reply aten/src/ATen/native/Activation.cpp]: Just making the speculation explicit is sufficient to convince clang to vectorize the loop where the naïve version did  not get vectorized.

Comment (weiyangfb) [reply aten/src/ATen/native/Activation.cpp]: Thanks a lot! Indeed I should've cached the input_data! I will compress the nested loops into just one loop later on

**Comment by weiyangfb** [comment]
@ezyang I will keep the THNN & THCUNN code for now since @cpuhrsch is planning to nuke /legacy/nn

File: aten/src/ATen/native/Activation.cpp [hunk 4]
```diff
@@ -43,6 +44,218 @@ Tensor & rrelu_(Tensor & self, Scalar lower, Scalar upper, bool training, Genera
...
+      prelu_cpu_backward_kernel_multi_weights<scalar_t>(
+        input,
+        weight,
+        grad_out,
+        input_grad,
+        weight_grad_collector,
+        input_stride0,
+        input_stride1);
+    });
+    // update weight_grad
```
Comment (weiyangfb) [inline aten/src/ATen/native/Activation.cpp]: is there a better way to do the reduction on weight? @SsnL

File: aten/src/ATen/native/cuda/Activation.cu [hunk 5]
```diff
@@ -2,9 +2,239 @@
...
+        input.data<scalar_t>(),
+        weight.data<scalar_t>(),
+        grad_out.data<scalar_t>(),
+        input_grad.data<scalar_t>(),
+        weight_grad_collector.data<scalar_t>(),
+        input_stride0,
+        input_stride1,
+        input_numel);
+    });
+    // update weight_grad
```
Comment (weiyangfb) [inline aten/src/ATen/native/cuda/Activation.cu]: is there a better way to do the reduction on weight? @SsnL

**Comment by cpuhrsch** [comment]
By "CPU_tensor_apply" you all are referring to TH? We have an equivalent [within ATen as well](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/CPUApplyUtils.h). Presumably that will be replaced by TensorIterator as well? @colesbury

**Comment by cpuhrsch** [comment]
In addition to this we also have [Parallel.h](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Parallel.h) that allows us to abstract the OMP calls.

File: aten/src/ATen/native/Activation.cpp [hunk 6]
```diff
@@ -43,6 +44,218 @@ Tensor & rrelu_(Tensor & self, Scalar lower, Scalar upper, bool training, Genera
...
+  int64_t i;
+  int64_t input_numel = input.numel();
+  auto result_data = result.data<scalar_t>();
+  auto input_data = input.data<scalar_t>();
+  auto weight_val = weight.data<scalar_t>()[0];
+
+  #pragma omp parallel for private(i) if (input_numel > 1000)
+  for (i = 0; i < input_numel; i++) {
+    scalar_t input_data_val = input_data[i];
+    result_data[i] = (input_data_val > 0) ? input_data_val : weight_val * input_data_val;
```
Comment (SsnL) [inline aten/src/ATen/native/Activation.cpp]: Be aware of https://github.com/pytorch/pytorch/pull/9206 . Looks like you are reverting an optimization here. Could you redo that and add a note?

File: aten/src/ATen/native/Activation.cpp [hunk 7]
```diff
@@ -43,6 +44,218 @@ Tensor & rrelu_(Tensor & self, Scalar lower, Scalar upper, bool training, Genera
...
+  int64_t i;
+  int64_t input_numel = input.numel();
+  auto result_data = result.data<scalar_t>();
+  auto input_data = input.data<scalar_t>();
+  auto weight_data = weight.data<scalar_t>();
+
+  // multiply values at each channel with weight[channel]
+  #pragma omp parallel for private(i) if (input_numel > 1000)
+  for (i = 0; i < input_numel; i++) {
+    int64_t channel = (i % input_stride0) / input_stride1;
```
Comment (SsnL) [inline aten/src/ATen/native/Activation.cpp]: This isn't similar to the original implementation at all. Indeed as @resistor said, modulo and division are expensive. It is possible that the original simple strategy of parallelizing over batch dim is better.

File: aten/src/ATen/native/Activation.cpp [hunk 8]
```diff
@@ -43,6 +44,218 @@ Tensor & rrelu_(Tensor & self, Scalar lower, Scalar upper, bool training, Genera
...
+  auto weight_val = weight.data<scalar_t>()[0];
+  auto grad_out_data = grad_out.data<scalar_t>();
+  auto input_grad_data = input_grad.data<scalar_t>();
+  auto weight_grad_data = weight_grad.data<scalar_t>();
+
+  #pragma omp parallel for private(i) reduction(+:sum) if (input_numel > 1000)
+  for (i = 0; i < input_numel; i++) {
+    scalar_t input_data_val = input_data[i];
+    scalar_t grad_out_data_val = grad_out_data[i];
+    input_grad_data[i] = (input_data_val > 0) ? grad_out_data_val : weight_val * grad_out_data_val;
```
Comment (SsnL) [inline aten/src/ATen/native/Activation.cpp]: This is optimizable too.

**Comment by cpuhrsch** [comment]
Another thing to (potentially) consider here is writing this in terms of vec256 within native/cpu for the CPU kernel - see https://github.com/pytorch/pytorch/pull/11565 as an example. This is much more involved however and mostly benefits CPU-bound kernels.

**Comment by cpuhrsch** [comment]
@resistor - the only issue is that this won't be compiled with -mavx/-mavx2 etc. if it doesn't live within native/cpu for the OSS version, where we rely on a dispatch to not assume avx/avx2 capabilities for any CPU. But we can easily move it into there and use the autovectorization. Having said that, just because the compiler vectorizes it doesn't mean it's faster, because it might try to maintain invariants we don't care about. For example, we might want to tradeoff adding one number at a time with adding 8 floats to 8 adders at a time. This won't be the same results, but it's much faster and will still be reproducible.

File: aten/src/ATen/native/Activation.cpp (hunk 6 above)
Comment (weiyangfb) [reply aten/src/ATen/native/Activation.cpp]: Very nice! Will change accordingly.

**Comment by resistor** [comment]
@cpuhrsch - Makes sense re: dispatch.

re: autovectorization, for this particular kernel there's not a lot of room for variation, since's it's a purely element-wise filter.  Getting the compiler to do the right thing for an accumulator loop is a lot trickier and more fragile.

File: aten/src/ATen/native/Activation.cpp (hunk 7 above)
Comment (resistor) [reply aten/src/ATen/native/Activation.cpp]: I suspect the best combination is to parallelize over the outermost dim and vectorize over the innermost dim.

**Comment by cpuhrsch** [comment]
@resistor - on element-wise operations we also have [vml.h](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cpu/vml.h) that defines a set of functions that can act as an interface to Intel's VML, or a loop + SLEEF if that's not available. Intel VML is much faster and also parallelizes (but of course doesn't let you merge operations). It's also worth checking whether some parts here can be gutted and replaced by Ops within MKL/MKL-DNN/iDeep etc. 

Having said this, in my mind, none of this is required for this and there's still value in dumping and simplifying code from TH/THNN and putting it into native.

File: aten/src/ATen/native/Activation.cpp (hunk 7 above)
Comment (resistor) [reply aten/src/ATen/native/Activation.cpp]: @cpuhrsch makes a good point that it's not clearly worth the effort to figure out the inner vectorization for this inline implementation.

**Comment by weiyangfb** [comment]
After making some changes on the CPU kernel (removing mod & div, split lines for compiler optimization, etc..), now the runtime looks much nicer:
```
>>> m = nn.PReLU()
>>> x = torch.randn(100, 100, 100, requires_grad=True)
>>> %timeit y = m(x)
1000 loops, best of 3: 223 µs per loop

>>> y = m(x).sum()
>>> %timeit y.backward(retain_graph=True)
100 loops, best of 3: 3.3 ms per loop

>>> m = nn.PReLU(100)
>>> x = torch.randn(100, 100, 100, requires_grad=True)
>>> %timeit y = m(x)
1000 loops, best of 3: 219 µs per loop

>>> y = m(x).sum()
>>> %timeit y.backward(retain_graph=True)
100 loops, best of 3: 3.27 ms per loop
```

**Comment by weiyangfb** [comment]
@cpuhrsch by `CPU_tensor_apply` I meant the one at ATen. At the beginning I had some performance with it, I guess it was mainly because all the optimization tricks weren't carried through. It would be nice to have `TensorIterator` here as well, but I haven't found an easy way to replace `CPU_tensor_apply` directly.

File: aten/src/ATen/native/cuda/Activation.cu [hunk 9]
```diff
@@ -2,9 +2,239 @@
...
+
+  AT_CHECK(input.is_contiguous());
+  AT_CHECK(weight.is_contiguous());
+
+  int64_t weight_num = weight.numel();
+  Tensor result = at::empty_like(input);
+  auto strides = input.strides();
+
+  // case1: shared weight for all channels
+  if (weight_num == 1) {
```
Comment (weiyangfb) [inline aten/src/ATen/native/cuda/Activation.cu]: CUDA kernel for weight.numel() == 1 is still slower than previous implementation, not sure why... can it be the difference between `THC_pointwiseApply2` and `CUDA_tensor_apply2`?

https://github.com/pytorch/pytorch/blob/b91b15d86e274611f448caa3f9b73eb4bf24a625/aten/src/THCUNN/generic/PReLU.cu#L17-L20

https://github.com/pytorch/pytorch/blob/b91b15d86e274611f448caa3f9b73eb4bf24a625/aten/src/THCUNN/PReLU.cu#L9-L23

File: aten/src/ATen/native/cuda/Activation.cu [hunk 10]
```diff
@@ -2,9 +2,239 @@
...
+template <typename scalar_t>
+void prelu_cuda_kernel_share_weights(
+  const Tensor& input,
+  Tensor& result,
+  const scalar_t* weight_data) {
+
+  at::cuda::CUDA_tensor_apply2<scalar_t, scalar_t>(
+    input,
+    result,
+    [=] __device__ (
```
Comment (SsnL) [inline aten/src/ATen/native/cuda/Activation.cu]: Not sure about the perf difference. But passing by reference might be faster consider that a pointer is pretty wide on GPU. You also try porting the exact same apply kernel over (just temporarily) and see if the problem is the apply fn.

Comment (weiyangfb) [reply aten/src/ATen/native/cuda/Activation.cu]: just realize devgpu ash and prn have different GPUs (K40m vs M40)... I was running the perf experiments on these two types of machines...I guess that's why there is a difference..

Comment (SsnL) [reply aten/src/ATen/native/cuda/Activation.cu]: haha I see :)

File: aten/src/ATen/native/Activation.cpp (hunk 1 above)
Comment (weiyangfb) [reply aten/src/ATen/native/Activation.cpp]: this is fixed by parallelizing the kernel

**Comment by weiyangfb** [comment]
Other than the CPU performance improvement with Vec256 (I probably can do it in a separate PR), do I need to do further changes on this PR? @ezyang

**Review by ezyang — APPROVED** [review]
Nope, thank you for debugging the perf issue, it is much appreciated.

**Pull request closed without merging**